[CH #57710] Interactivity popups

manmorjim opened the pull request:
### Changes

- Fixes issue: sometimes the feature cannot be clicked

- Fixes issue: Cannot disable hovers

- Add cursor pointer when the mouse is on a feature

- Add Airship popup component as Popup style

- Improves examples: Examples similar to VL: https://carto.com/developers/carto-vl/examples/#example-add-pop-ups-on-click
  - Feature hover
  - Feature click
  - Add pop-ups on hover
  - Add pop-ups on click
  - Interactive based styling

- Fixes issue: event error if layer was not added to the map

manmorjim committed:
Improve Popup examples like cartoVL

https://carto.com/developers/carto-vl/examples/#example-add-pop-ups-on-click

manmorjim committed:
New interaction API. Adapt popups to this new API

manmorjim committed:
Fix type in tests

manmorjim committed:
Add popup styles and adjust its position

manmorjim committed:
Update examples

manmorjim committed:
Refactor in Popup class

VictorVelarde commented on examples/visualization/interactions/add-popups-onclick.html:
```diff
@@ -0,0 +1,104 @@
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://libs.cartocdn.com/airship-style/v2.1.1/airship.css"
```
We can point to minor, and get patches for free, if we use a url like:
https://libs.cartocdn.com/airship-style/v2.1/airship.css (it also works with https://libs.cartocdn.com/airship-style/v2/airship.css)

VictorVelarde replied:
... and the last one is 2.3 (you can use minified versions too, like airship.min.css). In fact, I think I'm gonna add the lib to 'template' for other examples

manmorjim committed:
Add interactive based styling: hover & click

manmorjim committed:
Update interactive based styling example

manmorjim committed:
Merge branch 'web-sdk' into mmorillo/ch57710/interactivity-popups

manmorjim committed:
Adapt examples to new structure

manmorjim committed:
Merge branch 'web-sdk' into mmorillo/ch57710/interactivity-popups

alasarr reviewed (changes requested):
It looks great and it works great! Congrats on the nice work.

As an improvement, I feel that Layer.ts is growing too much. Could you try to move some logic of popup to an external module? 

Something like LayerInteration.ts where you can encapsulate all the code related to the layer and the interactivity, so you can define there the code of layer.on, _setPopupHandler, setPopupHover, setPopupClick, _wrapInteractiveStyle

VictorVelarde commented on examples/interactivity/add-popups-onclick.html:
```diff
@@ -0,0 +1,104 @@
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://carto.com/developers/carto-vl/v1.4.4/examples/maps/style.css"
```
We should remove this VL-thing. At the very end, it's equivalent to our map 100% custom style

VictorVelarde commented on examples/interactivity/add-popups-onclick.html:
```diff
@@ -0,0 +1,104 @@
+  <head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>GeoJSON Layer with CARTO Tiles Example</title>
```
Not urgent, but we should update titles too

VictorVelarde commented on examples/interactivity/add-popups-onclick.html:
```diff
@@ -0,0 +1,104 @@
+        height: 100vh;
+      }
+
+      div.as-infowindow > div.as-infowindow__content > button.as-btn {
+        float: right;
+        margin-top: -10px;
+        margin-right: -10px;
+      }
```
This is ok for now. I will try to update the styles for examples in the near future, to use airship all around (so styling is homogeneous with popups)

VictorVelarde commented on examples/interactivity/add-popups-onhover.html:
```diff
@@ -0,0 +1,94 @@
+          { attr: 'pop_2015', title: 'Population' }
+        ]);
+        await populationLayer.addTo(deckMap);
+        window.deckMap = deckMap;
```
I would remove the window.deckMap if not required. This is nice for debugging, but we should try to be homogenous on examples. These examples will be a very relevant part of the documentation!

manmorjim committed:
Refactor Layer using the class LayerInteractivity

- Created the class `LayerInteractivity` to manage all the layer
interactions like events and popups.

- `Layer` class moved to a folder called `layer` which contains also:
  - LayerInteractivity class
  - LayerOptions

- Updated Layer imports with the new location of the class.

VictorVelarde commented on packages/viz/src/lib/layer/Layer.ts:
```diff
@@ -0,0 +1,273 @@
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private _deckLayer?: any;
+
+  private _interacitivity: LayerInteractivity;
```
👀 typo: `_interacitivity`

VictorVelarde commented on packages/viz/src/lib/layer/Layer.ts:
```diff
@@ -0,0 +1,273 @@
+    this._source = buildSource(source);
+    this._style = buildStyle(style);
+
+    let hoverStyle;
+
+    if (options && options.hoverStyle) {
+      hoverStyle = buildStyle(options.hoverStyle);
+    }
+
+    let clickStyle;
+
+    if (options && options.clickStyle) {
+      clickStyle = buildStyle(options.clickStyle);
+    }
+
+    this._interacitivity = new LayerInteractivity(
+      this,
+      this.getStyle.bind(this),
+      this.setStyle.bind(this),
+      hoverStyle,
+      clickStyle
+    );
```
How about a small refactor, having a helper method to do `this._interactivity = buildInteractivity(options)`?

manmorjim committed:
Update example titles

VictorVelarde commented on packages/viz/src/lib/layer/Layer.ts:
```diff
@@ -22,19 +28,38 @@ export class Layer implements StyledLayer {
+    );
     this._options = {
       id: `${this._source.id}-${Date.now()}`,
+      ...this._interacitivity.getDefaultOptions(),
```
typo here

manmorjim committed:
Changes from CR

manmorjim committed:
Using interactivity props in the same way than styles

manmorjim committed:
Merge branch 'web-sdk' into mmorillo/ch57710/interactivity-popups

alasarr commented on packages/viz/src/lib/layer/Layer.ts:
```diff
@@ -201,51 +227,15 @@ export class Layer implements StyledLayer {
-
-      this._options.onClick = undefined;
-    }
+    this._interactivity.setPopupClick(elements);
```
Vamos!! 👌
